src/libopensc/card-cardos5.c

```diff
@@ -0,0 +1,1746 @@
+	{ NULL, NULL, NULL, 0, 0, NULL }
+};
+
+typedef struct {
+	uint8_t	*ptr;
+	size_t	 size;
+	size_t	 bytes_used;
+} buf_t;
+
+typedef struct {
+	uint8_t *encoded_ptr;
+	size_t	 encoded_len;
+	size_t	 raw_len;
+} coordinate_t;
+
+struct cardos5_private_data {
+	/* Current Security Environment Algorithm */
+	unsigned int	cse_algorithm;
+};
+
+static void
+buf_init(buf_t *buf, uint8_t *ptr, size_t size)
+{
+	buf->ptr = ptr;
+	buf->size = size;
+	buf->bytes_used = 0;
+}
+
+static int
+asn1_put_tag(uint8_t tag, const void *tag_content, size_t tag_content_len,
+    buf_t *buf)
+{
+	int r;
+
+	r = sc_asn1_put_tag(tag, (const uint8_t *)tag_content, tag_content_len,
+	    buf->ptr, buf->size - buf->bytes_used, &buf->ptr);
+	if (r == SC_SUCCESS) {
+		buf->bytes_used += tag_content_len + 2;
+		return 0;
+	}
+
+	return -1;
+}
+
+static int
+asn1_put_tag0(uint8_t tag, buf_t *buf)
+{
+	return asn1_put_tag(tag, NULL, 0, buf);
+}
+
+static int
+asn1_put_tag1(uint8_t tag, uint8_t tag_value, buf_t *buf)
+{
+	const uint8_t	tag_content[1] = { tag_value };
+
+	return asn1_put_tag(tag, tag_content, sizeof(tag_content), buf);
+}
+
+static int
+add_acl_tag(uint8_t am_byte, unsigned int ac, unsigned int key_ref, buf_t *buf)
+{
+	uint8_t	crt_buf[16];
+	buf_t	crt;
+
+	if (am_byte != 0xff)
+		if (asn1_put_tag1(ARL_ACCESS_MODE_BYTE_TAG, am_byte, buf))
+			return -1;
+
+	switch (ac) {
+	case SC_AC_NONE:
+		/* SC_AC_NONE means operation ALWAYS allowed. */
+		return asn1_put_tag0(ARL_ALWAYS_TAG, buf);
+	case SC_AC_NEVER:
+		return asn1_put_tag0(ARL_NEVER_TAG, buf);
+	case SC_AC_CHV:
+	case SC_AC_TERM:
+	case SC_AC_AUT:
+		if ((key_ref & BACKTRACK_BIT) || key_ref > UINT8_MAX)
+			return -1;
+
+		buf_init(&crt, crt_buf, sizeof(crt_buf));
+
+		if (asn1_put_tag1(CRT_TAG_PINREF, (uint8_t)key_ref, &crt) ||
+		    asn1_put_tag1(CRT_TAG_KUQ, KUQ_USER_AUTH, &crt) ||
+		    asn1_put_tag(ARL_USER_AUTH_TAG, crt_buf, crt.bytes_used,
+		    buf))
+			return -1;
+		return 0;
+	default:
+		return -1;
+	}
+}
+
+static int
+bertlv_put_tag(uint8_t tag, const uint8_t *data, size_t length, buf_t *buf)
+{
+	if (length > UINT16_MAX || buf->bytes_used == buf->size)
+		return -1;
+
+	*(buf->ptr)++ = tag;
+	buf->bytes_used++;
+
+	if (length < 0x80) {
+		if (buf->bytes_used == buf->size)
+			return -1;
+		*(buf->ptr)++ = (uint8_t)length;
+		buf->bytes_used++;
+	} else if (length < 0xFF) {
+		if (buf->size - buf->bytes_used < 2)
+			return -1;
+		*(buf->ptr)++ = 0x81;
+		*(buf->ptr)++ = (uint8_t)length;
+		buf->bytes_used += 2;
+	} else {
+		if (buf->size - buf->bytes_used < 3)
+			return -1;
+		*(buf->ptr)++ = 0x82;
+		*(buf->ptr)++ = (uint8_t)(length >> 8);
+		*(buf->ptr)++ = (uint8_t)(length & 0xFF);
+		buf->bytes_used += 3;
+	}
+
+	if (buf->bytes_used - buf->size < length)
+		return -1;
+
+	memcpy(buf->ptr, data, length);
+	buf->ptr += length;
+	buf->bytes_used += length;
+
+	return 0;
+}
+
+static int
+bertlv_get_tag(uint16_t tag, uint8_t **tag_content, uint16_t *tag_length,
```

Comment:
Imagine the following situation:
* It turns out that there *is* a problem with your implementation
* You are not available for having a deep look at it
* Others don't have a cardos 5 card available to look at it
In result, nobody can fix the issue and we are stuck with problems like memory leaks forever (see #395 for an examlpe). Unfortunately things like memory leaks *do* affect the whole library.

Others could look and test the generic functions, but not the stuff that is implemented in your card driver. So please, please don't reinvent the wheel and use what is already present.

Also, please note that, yes, you could in theory build a webbrowser in your card driver for some fancy reason (e.g. "it supports ASN.1"), but nobody wants that.